### PR TITLE
FEC: validate source block length

### PR DESCRIPTION
**Pull request by dshil**
Closes: #178

### CODE REVIEW

File: src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp [hunk 1]
```diff
@@ -71,6 +71,20 @@ bool OFDecoder::valid() const {
     return valid_;
 }
 
+size_t OFDecoder::max_source_block() const {
+    roc_panic_if_not(valid());
+
+    if (codec_id_ == OF_CODEC_REED_SOLOMON_GF_2_M_STABLE) {
+        return OF_REED_SOLOMON_MAX_NB_SOURCE_SYMBOLS_DEFAULT;
```
Comment (gavv) [inline src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp]: I've checked OpenFEC sources and it appears that these defines are just defaults (as the name suggests).

OpenFEC provides two function to get and set these and other codec parameters:

* of_get_control_parameter
* of_set_control_parameter

However, of_set_control_parameter() is not implemented for LDPC and RS currently.

I suggest to do the following:

* use of_get_control_parameter() instead of defines;

* create an issue (in Roc) for:
  * crating a patch for OpenFEC that implements of_set_control_parameter()
  * using of_set_control_parameter() in Roc sender with the block sizes provided by user (to remove any limitation)
  * using of_set_control_parameter() in Roc receiver either with the block sizes provided by session (to remove any limitation) OR with the maximum block sizes set by user via the receiver config (to keep some limit possible but make it configurable)

File: src/modules/roc_fec/reader.cpp [hunk 2]
```diff
@@ -114,6 +114,15 @@ bool Reader::update_size_(size_t new_sblen) {
         return true;
     }
 
+    if (new_sblen > decoder_.max_source_block()) {
```
Comment (gavv) [inline src/modules/roc_fec/reader.cpp]: The documentation for max_source_block() states that it can return zero, but there is no check here. We should either fix the documentation or the usage.

If we will use of_get_control_parameter(), we probably can just panic on its error (it should not happen) and fix documentation.

File: src/modules/roc_fec/writer.h [hunk 3]
```diff
@@ -55,7 +55,7 @@ class Writer : public packet::IWriter, public core::NonCopyable<> {
     bool valid() const;
 
     //! Set number of source packets per block.
-    void resize(size_t sblen);
+    bool resize(size_t sblen);
```
Comment (gavv) [inline src/modules/roc_fec/writer.h]: We should also replace `resize(n)` with `CHECK(resize(n))` everywhere in tests.

File: src/tests/roc_fec/target_openfec/test_encoder_decoder.cpp [hunk 4]
```diff
@@ -182,5 +191,10 @@ TEST(encoder_decoder, load_test) {
     }
 }
 
+TEST(encoder_decoder, max_source_block) {
+    check_max_source_block(ReedSolomon8m, OF_REED_SOLOMON_MAX_NB_SOURCE_SYMBOLS_DEFAULT);
+    check_max_source_block(LDPCStaircase, OF_LDPC_STAIRCASE_MAX_NB_SOURCE_SYMBOLS_DEFAULT);
```
Comment (gavv) [inline src/tests/roc_fec/target_openfec/test_encoder_decoder.cpp]: Other encoder-decoder tests use a cycle over all codecs:

```
for (int type = ReedSolomon8m; type != CodecTypeMax; ++type) {
```

It's better to use it here as well. And thus we can also remove (inline) the check_max_source_block() method.

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp [hunk 5]
```diff
@@ -1043,5 +1154,22 @@ TEST(writer_reader, writer_resize_block) {
     }
 }
 
+TEST(writer_reader, source_block_max) {
+    check_writer_source_block_max(rs_source_composer, rs_repair_composer, ReedSolomon8m,
```
Comment (gavv) [inline src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: I suggest to use the same approach as in encoder-decoder tests:

* create a cycle over all codec types in every test;
* inline test code (check_writer_source_block_max) in the test itself.

Pros:

* will automatically work with new codec if we add them;
* a single test-case code is not dispersed in different parts of the file.

File: src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp (hunk 1 above)
Comment (gavv) [reply src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp]: Discussed that we still need to validate that block size fits into FECFRAME packet limitations (1 or 2 bytes for ESI) so we still need a check. Leaving it as is for now, we'll add of_set_control_parameter() support later.

File: src/modules/roc_fec/idecoder.h [hunk 6]
```diff
@@ -23,14 +23,31 @@ class IDecoder {
 public:
     virtual ~IDecoder();
 
+    //! Get the maximum value of a source block for a used scheme.
+    //!
+    //! @returns
+    //!  0 when there are no limits.
+    virtual size_t max_source_block() const = 0;
```
Comment (gavv) [inline src/modules/roc_fec/idecoder.h]: Discussed to replace it with max_block_length().

File: src/modules/roc_fec/writer.h (hunk 3 above)
Comment (dshil) [reply src/modules/roc_fec/writer.h]: Done

File: src/modules/roc_fec/idecoder.h (hunk 6 above)
Comment (dshil) [reply src/modules/roc_fec/idecoder.h]: Done

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp (hunk 5 above)
Comment (dshil) [reply src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: Done

File: src/tests/roc_fec/target_openfec/test_encoder_decoder.cpp (hunk 4 above)
Comment (dshil) [reply src/tests/roc_fec/target_openfec/test_encoder_decoder.cpp]: Done

File: src/modules/roc_fec/reader.cpp (hunk 2 above)
Comment (dshil) [reply src/modules/roc_fec/reader.cpp]: Reworked.

File: src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp (hunk 1 above)
Comment (gavv) [reply src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp]: #193

File: src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp [hunk 7]
```diff
@@ -71,6 +71,16 @@ bool OFDecoder::valid() const {
...
 }
 
+size_t OFDecoder::max_block_length() const {
+    roc_panic_if_not(valid());
+
+    if (codec_id_ == OF_CODEC_REED_SOLOMON_GF_2_M_STABLE) {
+        return OF_REED_SOLOMON_MAX_NB_SOURCE_SYMBOLS_DEFAULT;
+    }
+
+    return OF_LDPC_STAIRCASE_MAX_NB_SOURCE_SYMBOLS_DEFAULT;
```
Comment (gavv) [inline src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp]: I suggest a switch with a panic in default branch, so that we notice the problem if we add a new FEC scheme ad forget to patch this method.

File: src/modules/roc_fec/target_openfec/roc_fec/of_encoder.h [hunk 8]
```diff
@@ -52,6 +52,9 @@ class OFEncoder : public IEncoder, public core::NonCopyable<> {
     //! Get buffer alignment requirement.
     virtual size_t alignment() const;
 
+    //! Get the maximum value of a source block for a used scheme.
+    virtual size_t max_block_length() const;
```
Comment (gavv) [inline src/modules/roc_fec/target_openfec/roc_fec/of_encoder.h]: Comment is outdated.

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp [hunk 9]
```diff
@@ -1653,5 +1653,129 @@ TEST(writer_reader, error_reader_decode_packet) {
...
+                      repair_composer(), packet_pool, buffer_pool, allocator);
+
+        Reader reader(config, decoder, dispatcher.source_reader(),
+                      dispatcher.repair_reader(), rtp_parser, packet_pool, allocator);
+
+        CHECK(writer.valid());
+        CHECK(reader.valid());
+
+        // try to resize writer with an invalid value
+        CHECK(!writer.resize(encoder.max_block_length() * 2));
```
Comment (gavv) [inline src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: I suggest to try `encoder.max_block_length() + 1` instead to catch off-by-one errors.

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp [hunk 10]
```diff
@@ -1653,5 +1653,129 @@ TEST(writer_reader, error_reader_decode_packet) {
...
+        // ensure that the block size was not updated
+        fill_all_packets(0);
+
+        // write packets to dispatcher
+        for (size_t i = 0; i < NumSourcePackets; ++i) {
+            writer.write(source_packets[i]);
+        }
+
+        // deliver packets from dispatcher to reader
+        dispatcher.push_written();
```
Comment (gavv) [inline src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: Let's check queue sizes before calling dispatcher to ensure that writer generated a block of correct size:

```
        CHECK(dispatcher.source_size() == NumSourcePackets);
        CHECK(dispatcher.repair_size() == NumRepairPackets);
```

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp [hunk 11]
```diff
@@ -1653,5 +1653,129 @@ TEST(writer_reader, error_reader_decode_packet) {
...
+
+        // write packets from queue to dispatcher
+        for (size_t i = 0; i < NumSourcePackets + NumRepairPackets; ++i) {
+            packet::PacketPtr p = queue.read();
+            CHECK(p);
+            if (i == 5) {
+                // violates: SBL + RBL > MAX_NES (for source packets)
+                p->fec()->source_block_length = encoder.max_block_length() * 2;
+                // reset block length to ensure that our packet won't be dropped
+                p->fec()->block_length = 0;
```
Comment (gavv) [inline src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: I think we need a test for both cases when block_length is zero and non-zero.

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp [hunk 12]
```diff
@@ -1653,5 +1653,129 @@ TEST(writer_reader, error_reader_decode_packet) {
...
+            }
+            if (i == NumSourcePackets + 3) {
+                // violates: SBL + RBL > MAX_NES (for repair packets)
```
Comment (gavv) [inline src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: `SBL + RBL <= MAX_NES`

also I think you need to remove " (for repair packets)"

File: src/modules/roc_fec/target_openfec/roc_fec/of_encoder.h (hunk 8 above)
Comment (dshil) [reply src/modules/roc_fec/target_openfec/roc_fec/of_encoder.h]: s/for a used scheme/for a scheme being used/

Comment (dshil) [reply src/modules/roc_fec/target_openfec/roc_fec/of_encoder.h]: Done

File: src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp (hunk 7 above)
Comment (dshil) [reply src/modules/roc_fec/target_openfec/roc_fec/of_decoder.cpp]: Done

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp (hunk 9 above)
Comment (dshil) [reply src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: Done

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp (hunk 10 above)
Comment (dshil) [reply src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: Done. Also additional block is processed.

File: src/tests/roc_fec/target_openfec/test_writer_reader.cpp (hunk 12 above)
Comment (dshil) [reply src/tests/roc_fec/target_openfec/test_writer_reader.cpp]: Done
